apps/zbugs/src/hooks/use-watch-query.tsx

```diff
@@ -0,0 +1,151 @@
+import type {Query, QueryType, Row, TableSchema} from '@rocicorp/zero';
+import {useEffect} from 'react';
+import {unreachable} from 'shared/src/asserts.js';
+import type {Change as IVMChange} from 'zql/src/ivm/change.js';
+import type {Input, Output} from 'zql/src/ivm/operator.js';
+import type {Format} from 'zql/src/ivm/view.js';
+import type {AdvancedQuery} from 'zql/src/query/query-internal.js';
+
+export function useWatchQuery<
+  TSchema extends TableSchema,
+  TReturn extends QueryType,
+>(
+  q: Query<TSchema, TReturn>,
+  onChange: (change: Change<TSchema>) => void,
+  enabled = true,
+): void {
+  useEffect(() => {
+    if (!enabled) {
+      return;
+    }
+    return (q as AdvancedQuery<TSchema, TReturn>)
+      .materialize(changeViewFactory)
+      .subscribe(onChange);
+  }, [q, onChange, enabled]);
+}
+
+class WatchQuery<TSchema extends TableSchema> implements Output {
+  readonly #subscribers = new Set<Callback<TSchema>>();
+  readonly #onDestroy: () => void;
+  #queryComplete = false;
+  #pendingChanges: IVMChange[] = [];
+
+  constructor(
+    input: Input,
+    onDestroy: () => void = () => void 0,
+    queryComplete: true | Promise<true>,
+  ) {
+    this.#onDestroy = onDestroy;
+    if (queryComplete === true) {
+      this.#queryComplete = true;
+    } else {
+      queryComplete.then(() => {
+        this.#queryComplete = true;
+        this.flush();
+      });
+    }
+    input.setOutput(this);
+  }
+
+  destroy() {
+    this.#onDestroy();
+  }
+
+  push(change: IVMChange): void {
+    this.#pendingChanges.push(change);
+  }
+
+  subscribe(cb: Callback<TSchema>): () => void {
+    this.#subscribers.add(cb);
+    return () => {
+      this.#subscribers.delete(cb);
+    };
+  }
+
+  flush() {
+    if (!this.#queryComplete) {
+      return;
+    }
+
+    for (const change of this.#pendingChanges) {
+      for (const cb of this.#subscribers) {
+        cb(toChange(change));
+      }
+    }
+    this.#pendingChanges = [];
+  }
+}
+
+function changeViewFactory<
```

Comment:
Hmmm I wonder if we can reasonably push this down a layer so that it can be used in non-React environments like Cloudflare-DO, NodeJS, vanilla JS.